### ISSUE TITLE
How does build order flag find the dependencies?

### ISSUE THREAD

**Issue by AliAskari** [conan-io/conan#1406]
I am trying to use the new `--build_order` flag in our CI but I am confused. I have a library without any dependency but all my other libraries depend on it. Assuming this is the reference for this library:

    UTIL/2.4.1@Ali/stable

I am expecting when I run `conan info --build_order UTIL/2.4.1@Ali/stable -j bo.json` to get a list of libraries , which in my case is almost all other libraries, that should potentially be rebuilt if this library is updated. Is that right? The generated json file is just an empty list

    {"groups": []}

However, if I go to a project that uses this library and run the same command I get a list of all libraries that should be updated and the order makes sense. I would like to know how does conan figure out this dependency. Are there any messages between the client and the server that determines which other libraries depend on this library?

The use case I see for this flag in our CI in case of a library update is to:

 - Find all dependent libraries downstream 
 - Determine which ones should be rebuilt
- Trigger their respective pipelines to test the new update 

but what I am seeing here is just a parsed output of the dependency graph, `--graph` flag. Although this is useful but I don't see how this flag could be useful for the CI.

**Comment by memsharded** [comment]
I think there might be a small misunderstanding. The ``--build_order`` does not provide a full inverse graph of all the possible dependencies in a remote that should be re-built. That is typically too much, and out of scope of a build. What you provide is two things:
- The consumer/root/project defining the dependency graph (in this case it is implicit in the default conanfile existing in the current directory)
- The package in the dependency graph that has changed (the reference ``UTIL/2.4.1@Ali/stable``)

What it provides is: given the dependency graph defined by the consumer conanfile, if the package reference ``UTIL/2.4.1@....```changes, which packages of the dependency graph have to be re-built

Does this make sense? Cheers!

**Comment by memsharded** [comment]
This question was answered long time ago. Some things have changed, and now (Conan 1.29) it is more recommended to use lockfiles and ``conan lock build-order`` command, using the right ``--build`` and ``package_id_mode`` to achieve the desired behavior, because it is quite necessary to apply a lockfile in order to build the packages in a graph separately or in parallel.
